Add usage and design documentation

eric-czech opened the pull request:
https://github.com/pystatgen/sgkit/issues/87

This adds documentation on data structures, API usage, missing data, pipelining and a few other topics.  

The changes here are mirrored at https://eric-czech.github.io/sgkit/ so that they're a little easier to review. 

I added an "Overview" section within "Getting Started" that introduces a bunch of these ideas without much depth.  I was imagining that this plus as separate "User Guide" section that explains some of these topics, plus others that are more specific to a certain type of genetic analysis, in greater detail and with realistic data would be a good start.  I went that direction mostly because it's how the Xarray documentation is structured.

@tomwhite I'm happy to move this over to the "Usage" section instead and omit any kind of less-detailed overview.  It may become a pain to maintain separate sections with some overlap.  I do have a slight preference for the name "User Guide" over "Usage" though -- let me know if you disagree.

@hammer my thinking with this is that the current design is more or less implicit in the user docs, but to capture some of the history on design decisions too I added a section in Contributing called "Design Decisions".  I put a few major threads I could think of in there.

FYI this also adds the IPython sphinx extension and ipython/matplotlib as doc build dependencies.

jeromekelleher commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,296 @@
+
+This example shows how either can be used, though users should prefer the mask array where possible since
+its on-disk representation is typically far smaller after compression is applied.
+
```
This gets hard-core pretty quickly. I wonder if we should put in an example where missing data is handled automatically by the library first? Users might get the impression they need to know what numba and jitting are to use this.

eric-czech replied:
https://github.com/pystatgen/sgkit/pull/278/commits/ae349eea8e8b2094f7d55bf282e5c15d74c4a84b#diff-9bc8bf6e8d9db020cefce1fbc0426795R196

eric-czech replied:
FYI: https://github.com/pystatgen/sgkit/pull/278#issuecomment-700997663

jeromekelleher commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,296 @@
+        # Compute Fst between the groups
+        # TODO: Refactor based on https://github.com/pystatgen/sgkit/pull/260
+        .pipe(lambda ds: sg.Fst(*(g[1] for g in ds.groupby('sample_cohort'))))
+        # Extract the single Fst value from the resulting array
```
This will change once Fst returns the full array of per-variant values by default, too.

eric-czech replied:
ok

jeromekelleher commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,296 @@
+Chaining operations
+~~~~~~~~~~~~~~~~~~~
+
+This example shows to chain multiple sgkit, xarray, and pandas operations into a single pipeline:
```
Is there some upstream documentation/tutorial explaining what an xarray/pandas pipeline is? I think this would be helpful here.

eric-czech replied:
https://github.com/pystatgen/sgkit/commit/ae349eea8e8b2094f7d55bf282e5c15d74c4a84b#diff-9bc8bf6e8d9db020cefce1fbc0426795R265

jeromekelleher commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,296 @@
+~~~~~~~~~~~~~~
+
+Chunked arrays, via Dask, operate very similarly to in-memory arrays within Xarray. Because of this, few affordances
+in sgkit are provided to treat them differently. They can generally be used in whatever context in-memory arrays are
```
I don't think the second sentence in this para adds much, probably best deleted.

eric-czech replied:
https://github.com/pystatgen/sgkit/commit/ae349eea8e8b2094f7d55bf282e5c15d74c4a84b#diff-9bc8bf6e8d9db020cefce1fbc0426795R305

jeromekelleher commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,296 @@
+
+Chunked arrays
+~~~~~~~~~~~~~~
+
```
A quick intro on what chunked arrays are with links to upstream docs would be super helpful here.

eric-czech replied:
https://github.com/pystatgen/sgkit/pull/278/commits/ae349eea8e8b2094f7d55bf282e5c15d74c4a84b#diff-9bc8bf6e8d9db020cefce1fbc0426795R305

jeromekelleher reviewed (commented):
This is excellent @eric-czech, thanks very much for taking the time to lay this all out. I've found it very helpful.

My main feedback is that it currently assumes a fairly advanced knowledge of the pydata ecosystem, and a few sentences here and there explaining high-level things and providing links to upstream docs would be a big help for many people (me included!)

eric-czech commented on docs/index.rst:
```diff
@@ -1,10 +1,23 @@
+down to smaller datasets and access simpler functionality for those that may be new to Python (though there is still
+a good bit of work to done on this front). See :ref:`getting_started` for more details.
+
+Sgkit is inspired heavily by `scikit-alelle <https://scikit-allel.readthedocs.io/en/stable/>`_ and `hail <https://hail.is/docs/0.2/index.html>`_,
```
Bah, thank you -- I cannot help myself with that e.

eric-czech commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,296 @@
+
+Overview
+--------
+
```
Think it should repeat what's on the [index page](https://eric-czech.github.io/sgkit/#sgkit-statistical-genetics-toolkit-in-python)?  Wasn't sure if you saw that, but it seems to be what you're describing.

jeromekelleher replied:
I did see the index page all right, but I still think it'd be worth giving a little context here. We should assume new users who have just clicked a link are coming in fresh here, so just a small bit of background would help get them started.

eric-czech replied:
https://github.com/pystatgen/sgkit/pull/278/commits/ae349eea8e8b2094f7d55bf282e5c15d74c4a84b#diff-9bc8bf6e8d9db020cefce1fbc0426795R25

eric-czech commented:
In https://github.com/pystatgen/sgkit/commit/ae349eea8e8b2094f7d55bf282e5c15d74c4a84b#diff-b7dd01040721465bb47249dc10bb65f2R17, I'm also moving `simulate_genotype_call_dataset` into the top-level API (https://github.com/pystatgen/sgkit/issues/252).  I made the default seed fixed too -- I don't see any reason for it to not be reproducible by default.

eric-czech commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,296 @@
+
+The presence of a single-nucleotide variant (SNV) is indicated above by the ``call_genotype`` variable, which contains
+an integer value corresponding to the index of the associated allele present (i.e. index into the ``variant_allele`` variable)
+for a sample at a given locus and chromosome. Every sgkit variable has a set of fixed semantics like this. For more
```
https://github.com/pystatgen/sgkit/commit/ae349eea8e8b2094f7d55bf282e5c15d74c4a84b#diff-9bc8bf6e8d9db020cefce1fbc0426795R63

eric-czech commented:
Lastly, these changes now also include:

- Moving `simulate_genotype_call_dataset` into the top-level API (https://github.com/pystatgen/sgkit/issues/252)
- Changing the default seed in `simulate_genotype_call_dataset` to be fixed -- I don't see any reason for it to not be reproducible by default (Hail does that with all simulation functions and its a nice convenience).
- Forcing the push to gh-pages in the docs workflow 
- Moving the dataset merge docs to Overview and creating a User Guide section with some stubs for sections I thought would be appropriate there: https://eric-czech.github.io/sgkit/user_guide.html

Could you take another look when you get a chance @alimanfoo / @jeromekelleher?

tomwhite commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,394 @@
+    # If an existing variable would be re-defined, a warning is thrown
+    import warnings
+    ds = sg.count_variant_alleles(ds)
+    with warnings.catch_warnings(record=True) as w:
```
This makes it look like users have to catch warnings, but you're having to that here to avoid ipython failing (I think). Not sure if there's a better way of doing it - one option would be to have the commented out code that would fail. Or add a comment being explaining that you don't need to explicitly catch warnings.

eric-czech replied:
I was trying to avoid the failure but more so I wanted a way to have it actually render somewhere.  I couldn't find a way to have the directive do that.  What's the problem with expecting users to catch warnings if they want to rather than having them print somewhere?  I'm struggling to word a comment for it.  Mind wording one for me and I'll drop it in?

eric-czech replied:
FYI haven't forgotten about this but I opened https://github.com/pystatgen/sgkit/issues/288 to track it.  I plan on working on https://github.com/pystatgen/sgkit/issues/287 soon so it may naturally go away.

tomwhite commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,394 @@
+`Method chaining <https://tomaugspurger.github.io/method-chaining.html>`_ is a common practice with Python
+data tools that improves code readability and reduces the probability of introducing accidental namespace collisions.
+Sgkit functions are compatible with this idiom by default and this example shows to use it in conjunction with
+Xarray and Xandas operations in a single pipeline:
```
Cool new project? :smile:

jeromekelleher replied:
Still have a "Xandas" typo here.

eric-czech replied:
🤦 ty.  I read @tomwhite's comment twice and still couldn't see the problem for some reason.  Xandas does sound cool though.

https://github.com/pystatgen/sgkit/pull/278/commits/dcb78dba17a0b78aa7cef5bf118253dadc0f881f

tomwhite commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,394 @@
+    # Show statistics for one of the arrays to be used as a filter
+    ds_qc.variant_call_rate.to_series().describe()
+
+    # Build a pipeline that filters on call rate and computes Fst between two populations
```
Maybe add "(or cohorts)" to echo the usage below.

eric-czech replied:
https://github.com/pystatgen/sgkit/commit/f7f5a039412fc6b4c29621d283c1a5738b746521#diff-9bc8bf6e8d9db020cefce1fbc0426795R268

tomwhite commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,394 @@
+- `Dataset.compute <http://xarray.pydata.org/en/stable/generated/xarray.Dataset.compute.html>`_ is called
+- `DataArray.compute <http://xarray.pydata.org/en/stable/generated/xarray.DataArray.compute.html>`_ is called
+- The ``DataArray.values`` attribute is referenced
+- Individual dask arrays are retrieved through the ``DataArray.data`` attribute and forced to evaluate via `Client.compute <https://distributed.dask.org/en/latest/api.html#distributed.Client.compute>`_, `dask.array.Array.compute <https://tutorial.dask.org/03_array.html#Example>`_ or by coercing them to another array type (e.g. using np.asarray)
```
Quote `np.asarray` and possibly link to its doc.

eric-czech replied:
https://github.com/pystatgen/sgkit/commit/f7f5a039412fc6b4c29621d283c1a5738b746521#diff-9bc8bf6e8d9db020cefce1fbc0426795R303

tomwhite commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,394 @@
+    sg.count_variant_alleles(ds).variant_allele_count
+
+
+A primary design goal in sgkit is to facilitate ad-hoc analysis. There are many useful functions in
```
Nit: "ad hoc"

eric-czech replied:
Fixed

tomwhite reviewed (approved):
Great work @eric-czech - thank you for writing it!

jeromekelleher commented:
Did you consider [jupyter-sphinx](https://jupyter-sphinx.readthedocs.io/en/latest/) for rendering the code chunks @eric-czech? Inspired by you, I've been trying to use the ``ipython-directive`` in another project and quickly hit against it's clunkiness. Jupyter-sphinx looks like it's much more active. Did you take a look at it?

eric-czech commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,394 @@
+
+- `Xarray <http://xarray.pydata.org/en/stable/>`_: N-D labeling for arrays and datasets
+- `Dask <https://docs.dask.org/en/latest/>`_: Parallel computing on chunked arrays
+- `Zarr <https://zarr.readthedocs.io/en/stable/>`_: Serialization for chunked arrays
```
https://github.com/pystatgen/sgkit/pull/278/commits/f7f5a039412fc6b4c29621d283c1a5738b746521#diff-9bc8bf6e8d9db020cefce1fbc0426795R31

eric-czech commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,394 @@
+
+There are currently no data models in the library that attempt to capture the complexity of many (or even common)
+analyses and the data structures that would support them -- operations are applied primarily to Xarray
+`Dataset <http://xarray.pydata.org/en/stable/data-structures.html#dataset>`_ objects instead. Users are free to manipulate data
```
This a little better? https://github.com/pystatgen/sgkit/pull/278/commits/f7f5a039412fc6b4c29621d283c1a5738b746521#diff-9bc8bf6e8d9db020cefce1fbc0426795R39-R42

eric-czech commented on docs/index.rst:
```diff
@@ -1,12 +1,25 @@
+down to smaller datasets and access simpler functionality for those that may be new to Python (though there is still
+a good bit of work to done on this front). See :ref:`getting_started` for more details.
+
+Sgkit is inspired heavily by `scikit-allel <https://scikit-allel.readthedocs.io/en/stable/>`_ and `hail <https://hail.is/docs/0.2/index.html>`_,
```
Fixed

eric-czech commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,394 @@
+`PLINK <https://www.cog-genomics.org/plink2>`_ or `BGEN <https://www.well.ox.ac.uk/~gav/bgen_format/>`_.
+This is a guideline however, and a ``Dataset`` seen in practice might include many more or fewer variables and dimensions.
+
+.. image:: _static/data-structures-xarray.jpg
```
Ah good call, I made it with google drawings so I made this read-only to anyone with the link so it could be duplicated+edited: https://github.com/pystatgen/sgkit/pull/278/commits/f7f5a039412fc6b4c29621d283c1a5738b746521#diff-9bc8bf6e8d9db020cefce1fbc0426795R46-R48

eric-czech commented:
> Maybe move the jit example to a section in the User Guide? It's not something most users need to get started, and could be quite intimidating.

Ok, I refactored it to this https://github.com/pystatgen/sgkit/commit/f7f5a039412fc6b4c29621d283c1a5738b746521#diff-9bc8bf6e8d9db020cefce1fbc0426795R209 and moved the numba example to a separate section in the user guide at https://github.com/pystatgen/sgkit/commit/f7f5a039412fc6b4c29621d283c1a5738b746521#diff-0fcaa96196be65c5d918fd5099ea19a8R50.  Sound good?

p.s. I don't know why git won't let me comment on the original thread inline even though it doesn't say it's outdated or otherwise inaccessible.

eric-czech commented:
> Did you consider jupyter-sphinx for rendering the code chunks @eric-czech? Inspired by you, I've been trying to use the ipython-directive in another project and quickly hit against it's clunkiness. Jupyter-sphinx looks like it's much more active. Did you take a look at it?

Ah nice, that does look better.  I opened https://github.com/pystatgen/sgkit/issues/287 to track the likely replacement.

eric-czech commented:
@jeromekelleher/@tomwhite do you want to take another look at this?  I'll leave it up to one of you to set the merge tag.

jeromekelleher reviewed (approved):
Looks great! I spotted one unfixed typo, but merge away whenever.

eric-czech commented:
Merged manually because this included a small change in the docs workflow.
